src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -490,37 +564,30 @@ private void InheritNullableStateOfFieldOrProperty(int targetContainerSlot, int
                 if (fieldOrPropertyType.IsNullable != false)
                 {
                     int targetMemberSlot = GetOrCreateSlot(fieldOrProperty, targetContainerSlot);
-                    if (targetMemberSlot >= this.State.Capacity) Normalize(ref this.State);
-
+                    bool? value = !fieldOrPropertyType.IsNullable;
                     if (isByRefTarget)
                     {
-                        // This is a property/field acesses through a by ref entity and it isn't considered declared as not-nullable. 
+                        // This is a property/field access through a by ref entity and it isn't considered declared as not-nullable. 
                         // Since reference can point to the heap, we cannot assume the property/field doesn't have null value after this assignment,
                         // regardless of what value is being assigned.
-                        this.State[targetMemberSlot] = (fieldOrPropertyType.IsNullable == true) ? (bool?)false : null;
                     }
                     else if (valueContainerSlot > 0)
                     {
                         int valueMemberSlot = VariableSlot(fieldOrProperty, valueContainerSlot);
-                        this.State[targetMemberSlot] = valueMemberSlot > 0 && valueMemberSlot < this.State.Capacity ?
-                            this.State[valueMemberSlot] :
-                            null;
+                        value = valueMemberSlot > 0 ? this.State[valueMemberSlot] : null;
                     }
-                    else
-                    {
-                        // No tracking information for the value. We need to fill tracking state for the target
-                        // with information inferred from the declaration. 
-                        Debug.Assert(fieldOrPropertyType.IsNullable != false);
 
-                        this.State[targetMemberSlot] = (fieldOrPropertyType.IsNullable == true) ? (bool?)false : null;
-                    }
+                    this.State[targetMemberSlot] = value;
                 }
 
-                if (fieldOrPropertyType.TypeSymbol.IsAnonymousType && fieldOrPropertyType.TypeSymbol.IsClassType())
+                if (valueContainerSlot > 0)
```

Comment:
>if (valueContainerSlot > 0) [](start = 16, length = 27)

This conditional logic, including the nested ```if``` doesn't feel right. I believe we should always inherit the state, either from slots, or from declaration information.

Reply:
Fields and properties of reference types are populated lazily, to avoid generating slots for members that are never referenced, and to handle cycles.